### ISSUE TITLE
port to SDL2

### ISSUE THREAD

**Issue by fabiangreffrath** [chocolate-doom/chocolate-doom#450]
Dear fraggle,

I have skimmed through the "SDL 1.2 to 2.0 Migration Guide" [1] and found one amazing feature that I believe fits ideal for Chocolate Doom's purpose. SDL2 allows for setting a "virtual resolution" for the game content and will take care of scaling that up to the native screen resolution itself. All this is done by the GPU using OpenGL if available or DirectX elsewhere. So, to me this sounds exactly like the direction the not-yet-ready opengl branch was heading to, but with an already established API.

Are there any plans to do the porting already?

 - Fabian

[1] https://wiki.libsdl.org/MigrationGuide

**Comment by fabiangreffrath** [comment]
Just a small (well, longer than intended) status update with regard to hardware-based rendering:

With the current branch HEAD in the sdl2-branch, hardware-based scaling and stretching does work with IMHO acceptable quality. The only drawback, but one that counts, is that it comes with a severe performance penalty! This, in turn, is caused by the many required rendering steps that our screen content has to go through: First, our paletted 8-bit framebuffer gets converted into an RGBA framebuffer of the same dimensions. This is then upscaled by a certain integer factor and blit into another framebuffer which is then uploaded into a texture and then scaled back to fit the screen or window. All of these steps, except for the last one, are done in software and thus cost CPU and RAM.

To mitigate this issue, I tried two experiments with mixed results:

1) Do all the colorspace conversion, scaling and stretching in one single step in software.
Inspired by the recent changes to the screenshot routines, my idea was to use the paletted 8-bit 320x200 screenbuffer and directly upload it into a texture pixel space of 1600x1200 dimensions (1600x1000 without aspect ratio correction). In the course of this, look up the RGB-value of each paletted pixel, add a 0xff alpha channel, and draw each pixel 5 times horizontally and either 5 or 6 times vertically, depending on aspect ration correction being enabled or not. Then, scale this texture to the screen by means of hardware rendering.
This approach worked surprisingly well -- once I understood what I was doing myself, that is. ;) But, it added back the same bottleneck of software-based pixel-wise scaling: The frame rates dropped down into the league of `chocolate-doom -geometry 1600x1200` -- which by hindsight should have been expected. The rendered picture, though, looked great.

2) The next idea was obvious: Do the intermediate bloating-up of the framebuffer in hardware as well.
The idea was to upload the RGBA framebuffer into an intermediate texture of original dimensions, render this into another texture of 1600x1200 dimensions and render the latter to screen. I won't get rid of the initial 8-bit palette -> RGBA colorspace conversion step with this, though.
For documentation purposes, this approach requires an additional intermediate texture that needs to get created with the `SDL_TEXTUREACCESS_STREAMING` flag, whereas the upscaled texture used for display needs the `SDL_TEXTUREACCESS_TARGET` flag. Then, we need a renderer that supports the `SDL_RENDERER_TARGETTEXTURE` flag.
Now comes the funny part: To move content from one texture into another, you don't *blit* between them, but you *render* -- and for this you need a *renderer*. Now, remember how setting `SDL_HINT_RENDER_SCALE_QUALITY` affects rendering quality if set from "nearest" to "linear"? Well, this setting applies *per renderer*. What we want to do, though, is render from the intermediate buffer into the upscaled one using the "nearest" algorithm and then render this to screen using the "linear" algorithm. This, however, is impossible using the same renderer (if I am wrong here, OMG, please tell me!).
So, one could think we just need a second renderer. In SDL (not sure if this is a general restriction with OpenGL, though), however, a renderer is directly bound to a window and you cannot have two renderers connected to the same window (again, if I am wrong here, please tell me!). Even, if the only thing the second renderer is supposed to do is render from one texture into another, *and not at all to screen*. So, we are left with either "linear" scaling in both steps which leads to a blurry, soft and washed-out picture -- or "nearest" scaling in both steps which creates the same artifacts in the status bar numbers as the initial implementation. 

Since opening a second empty program window just to satisfy this somehow random requirement is out of the question, this is where I am currently stuck. Any help is highly appreciated. The code is all there, but I am not going to push it in its current state.

TL;DR: I need a means to properly upscale a texture into another one using "nearest" algorithm whereas the second texture is supposed to get rendered to screen using "linear" algorithm.

**Comment by sneakernets** [comment]
Have we talked with Stella emulation team about this issue? They seemed to have solved this with the VCS.

**Comment by dendemann** [comment]
After successfully compiling on my raspberry pi using the SDL2 branch, the colors were screwed up. I fixed it by modifying I_SetPalette (i_video.c) and rearranging the color assignment from r,g,b to b,g,r. Is there a better way to fix this?

**Comment by fabiangreffrath** [comment]
Is this fixed by reverting https://github.com/chocolate-doom/chocolate-doom/commit/63fad45ab003f47fc5f28dd6f72e959e1d1a5f0d ? 

Doesn't SDL_UpdateTexture() already take care of this? It seems not so, sigh...

**Comment by fabiangreffrath** [comment]
> Have we talked with Stella emulation team about this issue? They seemed to have solved this with the VCS.

Are you sure they are tackling the same issue as we do, i.e. scaling up a paletted screenbuffer by an integer ratio and rendering it to screen with linear interpolation?

**Comment by dendemann** [comment]
> Is this fixed by reverting 63fad45 ?

Yes. Setting the pixel format to SDL_PIXELFORMAT_ARGB8888 works. The Pixel format returned by SDL_GetWindowPixelFormat(screen) is SDL_PIXELFORMAT_ABGR8888 on my pi.

**Comment by fabiangreffrath** [comment]
Alright, I have reverted this for now.
```
    TODO: This needs to be taken care of when creating the SDL_surfaces.
    At least we know that SDL_UpdateTexture() does not take care of the
    pixel order when uploading a SDL_surface to a texture.
```

**Comment by fabiangreffrath** [comment]
Hey @fragglet ,

I think we should take a decision on how to proceed with the rendering code.

Some observations: I agree that "nearest" scaling from the 320x200 surface to the texture makes the status bar font look clumsy for smaller screen geometries, e.g. 640x480. The current software scaling and stretching routines simply look better in these cases. The situation improves when introducing an intermediate upscaled surface for hardware rendering, but the performance penalties are significant. For higher screen geometries (e.g. 1600x1200), though, simple "nearest" hardware scaling does a good job and is nearly indistinguishable from the current software scaling, and should even outperform the former.

So, I suggest to do the following: Revert the code back to my initial implementation with only one 320x200 RGBA surface that is uploaded into the texture, i.e. no intermediate upscaled RGBA surface. Make hardware rendering strictly optional and bring back the software scaling routines. Add a menu item in `chocolate-doom-setup` to allow for choosing between both rendering methods and maybe give a hint that hardware-rendering is only recommended for bigger screen geometries, e.g. full screen with desktop resolution.

What do you think?

**Comment by J1mbo** [comment]
I did up scaling via textures on SDL1.2 directly from 320x240 frame buffer, which seemed to work just fine, what is the benefit of moving to SDL2 compared to that?

**Comment by fragglet** [comment]
Sorry I haven't responded yet. I need to re-read this whole thread and understand what the state is. If possible I'd really like to do hardware scaling through the SDL API (ie. I'd like to keep the work you've already done) but I need to look at the current code and understand what the problem(s) with it are.

**Comment by jmtd** [comment]
As an aside, I was looking over Fabian's work and saw some low-hanging fruit in `caf7461a` around setup, so I started to poke at that ([WIP branch](https://github.com/jmtd/chocolate-doom/tree/sdl2-jmtd-setup-simplify)). There's some more work to do on SDL input with `textscreen`, at least.

**Comment by EMue** [comment]
@fabiangreffrath

Please note: `SDL_HINT_RENDER_SCALE_QUALITY` applies per *texture*, not per renderer. The intermediate texture approach as outlined by you is definitely feasible.
IIRC the hint must be set before texture creation, but I'm not 100% certain about that. The order definitely matters.

Here's how I would do it:
1. Render to a 320x200 surface.
2. Do palette conversion in software and copy the result to a 320x200 texture.
3. Render into an intermediate texture using `nearest` integer scaling (e.g. to 640x400 or 960x600 - higher values look too harsh in my opinion).
4. Render to the window using `linear` scaling.

This completely avoids software scaling, and (optionally) softens pixel borders while still retaining the pixelated look. Please ask if you require sample code.

**Comment by fabiangreffrath** [comment]
@EMue Erik, this was the missing hint, the gold tip! Thank you so much!

It works! Everything is scaled and stretched and rendered by hardware and it looks great!

And have a look at this:

Chocolate Doom 2.2.0:
```
$ /usr/games/chocolate-doom -3 -timedemo demo3 | grep timed
timed 4471 gametics in 372 realtics (420.658600 fps)
```

Current sdl2-branch:
```
$ ./chocolate-doom -3 -timedemo demo3 | grep timed
timed 4471 gametics in 222 realtics (704.887390 fps)
```

I am close to tears. Will push my changes shortly!

**Comment by fragglet** [comment]
This is awesome news!! Thanks Fabian.

**Comment by sneakernets** [comment]
YES!

**Comment by bradharding** [comment]
Outstanding! And such a simple solution... Great work Fabian and @EMue!

**Comment by fragglet** [comment]
So I checked out the code. It looks good and I think this makes gl-branch obsolete. It's really good that we'll be able to do this entirely through the SDL2 API.

There's still some more stuff to do on sdl2-branch - most notably input-related. I think some of the keyboard inputs (eg. for typing in savegames) don't work if you have native keyboard input turned on. I need to look into that. The main problem is that it's hard to map to SDL2's new input scheme.

Re: scaling, I'm wondering what we should do about the software scaling code (i_scale.c). I know we've talked a bit about this in the past but we should probably figure out what to do. Now that we have this I'm kind of tempted to just delete it all and say this is how it's going to be done in the future. Modern hardware should all have hardware acceleration that makes this work well. The only cases where it's arguable that we should keep it are for older machines, but it's worth remembering that with SDL2 we're already losing a lot of support for older machines anyway. 

The original design goal of Chocolate Doom is to support modern hardware and with the change to SDL2 we're making a decision to essentially step into a new era. I think it might be time to leave the software scaling behind.

**Comment by fragglet** [comment]
So, having now tried out the new upscaler: it's a definite improvement. I have a couple of suggestions for how we can possibly make this even better.

We're scaling up from 320x200->640x400 and then using linear to scale that to the full screen. This works pretty well but there's still some noticeable blurriness. A scale factor of 3 might work better (or make it configurable?)

For gl-branch I was using an adaptive scaling selection - basically picking the smallest scale factor that resulted in a larger size than the target screen size. For example for 800x600 you scale to 960x600 and then scale down. Then you should always get "perfect blockiness". Note that the aspect ratio for the intermediate buffer doesn't necessarily have to match the original 320x200 so long as each side is an integer multiple in size. For example you could have an intermediate size of 640x800 and it doesn't matter so long as the final screen size you're drawing to is right.

Don't know if this is helpful.

**Comment by bradharding** [comment]
A couple of minor suggestions:
1) Since no clipping is done, you could use SDL_LowerBlit() instead of SDL_BlitSurface().
2) It might be worth using SDL_RenderClear() each frame, to make sure the pillarboxes are kept clear each frame.

**Comment by chungy** [comment]
> We're scaling up from 320x200->640x400 and then using linear to scale that to the full screen. This works pretty well but there's still some noticeable blurriness. A scale factor of 3 might work better (or make it configurable?)

I think the only thing that should be made certain is that people with video drivers capable of 320x200 or 640x400 should still look pixel-perfect, but this is probably not a real problem. If a factor of 3 works well, might as well standardize on it and call it good enough, no need for more configuration knobs.

**Comment by fragglet** [comment]
To clarify, if there are any configuration knobs they should be "secret" configuration options inside the .cfg file, something like this should not be exposed in the setup tool.

**Comment by chungy** [comment]
I just built the sdl2 branch, seems to work well, but on a 1920x1200 display, the scaling is very noticably blurry (on SDL1, Chocolate Doom pillarboxes a 1600x1200 image, which is absolutely crisp and pixel-perfect). It is very disappointing and even disorienting to look at. Changing UPSCALE to 6 makes it replicate the look of the original render. Perhaps this is a good reason for the secret configuration...

**Comment by fabiangreffrath** [comment]
>  Changing UPSCALE to 6 makes it replicate the look of the original render.

Well, there is quite some gap between 2 and 6. Does it already help if you increase UPSCALE to 3? Or does it help if you set the dimensions of texture_upscaled to 640x800, as suggested by @fragglet ?

> Is there hardware that wouldn't work with such a large one, perhaps platforms such as Raspberry Pi?

I could tell a difference, i.e. a small performance penalty, by increasing from 2 to 3. But I guess Intel on-board graphics on Linux isn't considered the most powerful system anyway. I bet that some modern nvidia or amd graphics will do the scaling near instantaneous.

**Comment by fabiangreffrath** [comment]
>  I think some of the keyboard inputs (eg. for typing in savegames) don't work if you have native keyboard input turned on.

I can confirm this. With my German keyboard I have to hit Z instead of Y to confirm the quit message and the mouse wheel does not work anymore.

> I think it might be time to leave the software scaling behind.

I feel the same. Plus, I believe that SDL2 still falls back to software scaling methods if hardware-based scaling is unavailable. I am not absolutely sure about this, though.

> For gl-branch I was using an adaptive scaling selection - basically picking the smallest scale factor that resulted in a larger size than the target screen size.

This means we would have to destroy and recreate the upscaled texture whenever we change the screen resolution e.g. by moving the window borders with the mouse. It's certainly doable, but maybe a bit awkward.

> To clarify, if there are any configuration knobs they should be "secret" configuration options inside the .cfg file, something like this should not be exposed in the setup tool.

Agreed. This is something that we (i.e. the devs) should find out, not the users.

**Comment by chungy** [comment]
> Well, there is quite some gap between 2 and 6. Does it already help if you increase UPSCALE to 3? Or does it help if you set the dimensions of texture_upscaled to 640x800, as suggested by @fragglet ?

I tried 3 and 4 before bumping to 6 (skipped 5). Both of them were still far too blurry for my taste. Still noticably blurry with 640x800, but less extreme.

**Comment by fabiangreffrath** [comment]
For aspect ratio corrected setups, the famous 5*SCREENWIDTH x 6*SCREENHEIGHT for the upscaled texture should be perfect.

**Comment by fabiangreffrath** [comment]
> For aspect ratio corrected setups, the famous 5xSCREENWIDTH x 6xSCREENHEIGHT for the upscaled texture should be perfect.

But this throws me back to half the fps ~330. Maybe introduce a user switch "quality vs. performance"? Phew...

**Comment by fragglet** [comment]
The aspect ratio of the intermediate texture doesn't matter.

**Comment by fabiangreffrath** [comment]
I believe it does. It decides if the aspect ratio correction is done in the first step by means of "nearest" scaling or in the second step by means of "linear" scaling.

**Comment by fragglet** [comment]
Nope. Only the final render to the screen determines the aspect ratio.

The intermediate texture can be something completely ridiculous like 320x1200 and it won't affect the final aspect ratio. The only thing it can affect is quality. So it's best to choose the intermediate texture in a way that creates the best quality rather than considering aspect ratio at all.

**Comment by jmtd** [comment]
But surely the aspect ratio of the intermediate texture has a bearing on the quality?

**Comment by fabiangreffrath** [comment]
That's what I was talking about.

**Comment by fragglet** [comment]
The size of the intermediate texture (ie. number of pixels) affects quality - its aspect ratio isn't the part that matters.

**Comment by fragglet** [comment]
I guess my real point is this: don't get hung up on the idea that the intermediate texture should necessarily have an aspect ratio that corresponds to the real screen - it's a false assumption. See my previous comment up above about the adaptive scaling approach that I was using for gl-branch. That might be a much better approach because the intermediate texture size is determined by the final screen size, removing the need for any kind of "quality slider" control.

**Comment by jmtd** [comment]
A little while ago I started poking at the setup code on this branch, I did [start cleaning out the fullscreen video modes](https://github.com/jmtd/chocolate-doom/commit/7639d1443e86f3e4f904883acf6564fb15008125) but got a bit hung up on the input issues, I might look at them next.

**Comment by fabiangreffrath** [comment]
Could someone with low-end hardware (e.g. Raspberry Pie) please check if rendering still works with the `upscale` factor set to 6? I think I have read somewhere that 1600x1200 is the maximum texture size that *must* be supported to still be considered OpenGL compliant. With a texture size of 1920x1200 we would break this barrier and would thus need to maintain separate `upscale` factors for vertical and horizontal upscaling.

**Comment by chungy** [comment]
If that's true, 1600x1200 wouldn't be an issue. It's still a pixel-perfect scaling of the 320x200 display.

I wonder if Raspberry Pi is even the lowest-end that should be tested... I'm sure there's plenty of Intel GPUs from 10+ years ago that would be worse :-)

**Comment by fabiangreffrath** [comment]
> It might be worth using SDL_RenderClear() each frame, to make sure the pillarboxes are kept clear each frame.

@bradharding I am not sure what you mean. Do you suggest do call SDL_RenderClear() in I_FinishUpdate()? I fail to see why. There isn't anything ever rendered into the pillarboxes and when the window is resized, SDL_RenderClear() is called anyway.

**Comment by bradharding** [comment]
> @bradharding I am not sure what you mean. Do you suggest do call SDL_RenderClear() in I_FinishUpdate()? I fail to see why. There isn't anything ever rendered into the pillarboxes and when the window is resized, SDL_RenderClear() is called anyway.

While playing in fullscreen, if you get something pop onto the screen (such as a general Windows 10 notification box, or Steam, or whatever) such that it draws over one of the pillarboxes, the pillarbox won't be cleared, and that part of the popup will remain there.

**Comment by fabiangreffrath** [comment]
Makes sense, thank you!

**Comment by dendemann** [comment]
I found the time to do a few benchmarks on the Raspberry Pi. The performance improvement is incredible as it allows smooth gameplay on the Pi now. Thank you guys and keep up the good work!

commit 63fad45
chocolate-doom -3 -timedemo demo3 | grep timed
timed 4471 gametics in 7594 realtics (20.606400 fps)

commit aafbdbb
chocolate-doom -3 -timedemo demo3 | grep timed
timed 4471 gametics in 3314 realtics (47.219372 fps)

**Comment by bradharding** [comment]
One thing I've noticed when using SDL2's linear filtering is that one side of the screen will "bleed" onto the other side. This is most noticeable in the automap, as you can see here along the left hand side in Doom Retro (it may help to view the image fullsize):

![image](https://cloud.githubusercontent.com/assets/3210148/9978613/4e5309aa-5f7d-11e5-9a33-32a2661b7dc5.png)

I've been working on a solution that effectively just paints a 1px wide black line either side of the screen to cover this bleeding up, but its a bit of a hack. Someone more familiar with SDL2 may be able to provide a better solution.

**Comment by bradharding** [comment]
It would seem `SDL_RenderSetClipRect()` is the function to use in this instance.

**Comment by fabiangreffrath** [comment]
Is it possible this is hardware-related? i cannot reproduce this bleeding, neither with Choco's sdl2-branch nor with Crispy's hicol2 branch.

**Comment by bradharding** [comment]
Hmm... The bleeding has been present for me on two different systems (my current PC with Radeon graphics, and previous one with Nvidia), and at various different resolutions. Can anybody else reproduce this?

**Comment by fabiangreffrath** [comment]
Does this only occur in the Automap? Only in rotate mode? Only for some specific types of lines?

**Comment by bradharding** [comment]
No, it's not just the automap. I just showed that as an example as it's easy to spot. It happens consistently, and is more noticeable if you have, say, a brighter area on one side as you move around. Here's the problem at player start in E1M1. You can see it on the left hand side, towards the middle:

![image](https://cloud.githubusercontent.com/assets/3210148/9979802/67b868a0-5fc2-11e5-9e42-36aaa4460271.png)

It's always been an issue for me in DR since starting with SDL2, but because I've used nearest-neighbor as default until recently, I haven't bothered looking into it.

**Comment by bradharding** [comment]
Fabian, I'm struggling to get the sdl2-branch of Chocolate Doom to compile using VS Community 2015 on my PC. Could you please post a link to a binary, and then I can confirm if I can replicate this issue? I don't think I'm doing (or not doing) anything different in DR that could be causing it...

**Comment by fabiangreffrath** [comment]
Oh, I am currently only building on Linux and have not yet built and installed SDL2 on my Windows VM.

**Comment by bradharding** [comment]
I think I've determined what the problem is.There's a bug with SDL2 when using Direct3D (see [here](https://bugzilla.libsdl.org/show_bug.cgi?id=1878)). I was wrong before when I said it occurred in OpenGL. SDL releases are few and far between, so I'll try the patch suggested in the bug report and compile it myself. It may be worth looking in to before releasing Chocolate Doom's SDL2 build.

**Comment by haleyjd** [comment]
This is a pretty common issue with hardware accelerated textures when filtering is involved. The usual solution is to have a border around the texture (so in other words, at some point, there is a black strip all the way around the choco framebuffer output), so that filtering functions do not pick up pixels off the other side of the texture ("wrap" mode) when calculating interpolated colors for pixels at the edge.

**Comment by fabiangreffrath** [comment]
Am Donnerstag, den 17.09.2015, 14:05 -0700 schrieb Jonathan Dowland:
> A little while ago I started poking at the setup code on this branch,
> I did start cleaning out the fullscreen video modes but got a bit
> hung up on the input issues, I might look at them next.

Thank you for this. However, I am a bit uneasy with outright removing
functioning code. Instead, I have so far commented out obsolete code by
using "#if 0" and comments. I believe it is safer that way until we
have finally decided which direction to go with regard to software
scaling in the SDL2 branch.

 - Fabian

**Comment by jmtd** [comment]
> Thank you for this. However, I am a bit uneasy with outright removing
> functioning code. Instead, I have so far commented out obsolete code by
> using "#if 0" and comments. I believe it is safer that way until we
> have finally decided which direction to go with regard to software
> scaling in the SDL2 branch.

As you wish. I personally prefer to use git history for that, but it's your branch, your rules :) (and the fact you left the scalers in actually helped me when I rebased [screenshots-properly](https://github.com/jmtd/chocolate-doom/tree/screenshots-properly-sdl2) onto sdl2-branch)

**Comment by fabiangreffrath** [comment]
Am Montag, den 21.09.2015, 06:50 -0700 schrieb Jonathan Dowland:
> As you wish. I personally prefer to use git history for that, but
> it's your branch, your rules :) (and the fact you left the scalers in
> actually helped me when I rebased screenshots-properly onto sdl2
> -branch)

Oops, this seems to be a misunderstanding: This is not my branch and I
am not making the rules! This was just a suggestion.

Actually, @fragglet started working on this branch and I was just
sliding in to implement hardware-based rendering. Now I consider this
feature mostly done. Adapting input (i.e. keyboard and mouse [wheel])
to the new API is still left to do and I believe that @fragglet is more
experienced in this field and has more expertise to do the actual
porting. After all, it is his project and I'll leave it to him to say
whether this branch is ready for prime time or not.

 - Fabian

**Comment by PorscheMonty** [comment]
Please don't do away with the software scaling functionality. It's served CD well for years and it so far appears to be the only way to get a pixel-perfect image on screen. 

Accuracy actually matters to some people (hello there...), and diminishing from it for the sake of being able to accommodate a 320x200 image on inferior flatpanel technology is insanity.

I use several 3:2 display modes ranging from 320x200 up to 1600x1000 (depending on the video hardware available to the computer) and perform aspect ratio correction directly from the monitor (all CRT's) and they all look and work perfectly fine with the current software scaler. I also use 1600x1200 occasionally when the video driver won't accept a custom display mode (normally 1600x1000) plus CD's aspect ratio correction functionality for a perfect 4:3 image.

If so be that the you don't consider the current scaler to be efficient enough for higher resolutions (CD, because of it's internal limitations, will crash long before it gets any opportunity to slow down anyways, even when outputting a 1600x1200 picture on a miserly 3ghz P4) just make it optional and give the user the opportunity to choose between a correct display (just as good ole' MAME does with it's archaic purist modes) and the blurry mess of compromise (yet reasonable to some extent) you're shooting for.

The DirectX video driver option was hidden from the setup and isn't even documented. Yet it's needed in order to get vsync on, and had I not used CD before, I wouldn't even know it was there! Please don't go down that road with this hardware scaling fad.

Very much appreciated.

**Comment by fragglet** [comment]
Thanks for the feedback. Suffice it to say that accuracy in how the scaling works matters very much to me (this is why I went to such pains to design the current software algorithm). Once the SDL2 port is complete, I believe deprecating / removing the old software code is the right thing to do, but I only say that because I'm confident that the new hardware scaler gives comparable (theoretically, it should even be superior) quality.

**Comment by AXDOOMER** [comment]
Will it be possible with the hardware scaling to still get the black borders around the screen? The software scaling puts black bars above and under the screen so the window through which we see the Doom world is always a multiple of 320. Else, the game has to do horizontal interpolation between the pixels and it makes the border around the each pixels very blurry. It's still very visible at 1920x1080 and it makes me feel uncomfortable. It destroys the sharpness of the pixels.

**Comment by chungy** [comment]
Aspect-correction with black bars is still available with hardware scaling. What shouldn't happen this time, however, is black bars on all four sides at once; in the software scaling modes, this is due to the rigidity of having only a fixed number of modes. Depending on the display's aspect ratio, hardware scaling should _only_ produce black bars on two sides at most (whenever the display is not 4:3). For most people, this will be on the left and right sides for a widescreen monitor. It's conceivable that some people will play with such monitors oriented in a portrait mode, and Chocolate Doom should render with black bars on the top and bottom instead.

**Comment by AXDOOMER** [comment]
On  a 16:9 display, Chocolate-Doom will scale it to the maximum vertically, so let's say my display is a 1080p, it will scale the game horizontally to 1440, but that's not a multiple of 320.

**Comment by chungy** [comment]
You already don't get what you're asking for with a 1920x1080 display with the software scaling.

**Comment by AXDOOMER** [comment]
Previously, 1920x1080 was scaling the screen texture to 1280x1000, but this has been fixed, now it's 1280x960.

**Comment by AXDOOMER** [comment]
Here's an illustration. I took screenshots and zoomed on them. The second picture is hardware scaling and you can see the deterioration of the sharpness of pixels. 


This is the result of the **software** scaling: (Vertical subpixels appeared, still looks good)
![vertical](https://cloud.githubusercontent.com/assets/6194072/10414678/f0ceac52-6fae-11e5-99f8-8c071bab5f01.png)



This is the result of the **hardware** scaling: (There are subpixels between each pixels, ugly)
![index](https://cloud.githubusercontent.com/assets/6194072/10414681/0db114d6-6faf-11e5-81d0-080c701e7536.png)

Closer look:
![inde2x](https://cloud.githubusercontent.com/assets/6194072/10414682/16e4933e-6faf-11e5-8297-320c3ebafae1.png)
This makes a use difference to my eye.

**Comment by PorscheMonty** [comment]
@fragglet: I'm very happy to hear there's a zeal for this type of accuracy on your part.

The good aspect of the whole deal, it will be extremely easy to spot differences between the output of both scalers, thus if the picture you get under a particular scaling factor in hardware mode doesn't match it's equivalent in software a 100% (1:1) down to the CRC32, we're indeed dealing with inaccuracy, and so should "quality" be equated with accuracy (in my humble, totally correct, unbiased opinion...), making a GPU's implementation (and there are many, and I have seen some weird stuff going on with SIS video hardware) of the necessary functionality, key to consistently getting a correct picture. Please be mindful of that as well.

As touching the non-scaled, 320x200 mode, will that remain untouched?
Regards.

**Comment by fabiangreffrath** [comment]
Maybe we could set the scaling mode to "nearest" instead of "linear" for very high resolutions? That is, for resolutions in which a duplicate pixel hurts less than an interpolated one (maybe starting from 3x or 4x upscaling).

**Comment by fragglet** [comment]
I think it probably makes sense to do some comparisons between the new scaling and old software scaling, but comparing it to the level of CRC32 is a bad idea.

When you do aspect ratio correction, unless you're at 1600x1200 there will be some interpolation between pixels taking place. The software scaler does this using 8-bit lookup tables that work well enough but are ultimately inaccurate. So the output from the software scaler shouldn't be considered a gold standard. This is one area where the hardware scaler can actually do better.

Alexandre-Xavier's point about inexact scale factors is quite interesting. With the software scaler you can only ever get blurriness in one axis. For example 640x480 is perfect horizontally but interpolated vertically; 800x600 is the opposite. With the hardware scaler you can potentially interpolate in both directions.

The doubling of the blurriness certainly has some effect. It's most noticeable at low resolutions - for example try running at something like 480x300 and you'll see. But I'm not convinced it's that big a deal at 1920x1080 where each pixel is already very fully overrepresented anyway.

We should probably add an option to behave like the current software scaler and always maintain integer scaling in one axis. But overall the benefit of being able to run at arbitrary resolutions, and particularly to be able to run properly fullscreen at resolutions like 1920x1080, isn't something we should just discard without seriously considering.

**Comment by AXDOOMER** [comment]
> Maybe we could set the scaling mode to "nearest" instead of "linear" for very high resolutions?

An option to set the scaling mode should be added in Chocolate-Setup. The scaling mode should not be dependent on the resolution, because how people see is different for everybody. The user should be allowed to chose what he wants. For example, I'd use the nearest filter at 1920x1080 because else I see the interpolation between the pixels and someone else would chose the linear filter because he doesn't mind about the blurriness or doesn't see it because he has a bad vision. 

> But I'm not convinced it's that big a deal at 1920x1080 where each pixel is already very fully overrepresented anyway.

Well, I see the blurriness between pixels at 1920x1080, so maybe I have goods eyes or a good screen. I'd prefer to play using the nearest filter than to see that there is interpolation between each pixels. 

> We should probably add an option to behave like the current software scaler and always maintain integer scaling in one axis.

That's a good idea.

**Comment by PorscheMonty** [comment]
Be it sharp or be it blurry, scaling error's still scaling error in my eyes, so aspect ratio-corrected modes concern me not in the least, as I normally only use 3:2 modes and perform the 4:3 stretching directly from the CRT.

1600x1200, however, is a completely different animal, because at that particular resolution the scaling will be even, that is, every line will be exactly as thick as the other.

As long a 3:2 modes and 1600x1200 specifically will pass the CRC32 test, I'm a happy man :)

**Comment by fragglet** [comment]
You'll be pleased to hear that 1600x1200 should be completely unchanged.

**Comment by fragglet** [comment]
I've made a gallery of the E1M1 start screen at three different resolutions: 640x480, 800x600 and 1280x960. For each, I took a screenshot at the actual resolution and then increased the window size by just a few pixels so it was almost the same but not exactly: this forces interpolation and allows us to compare the effect.

https://imgur.com/a/xZzvG

**Comment by PorscheMonty** [comment]
I'm going to put it as best as I know how, then will I shut up and shy away from your conversation.

Over the years I've grown very fond of Chocolate Doom, and I've used it and abused on a variety of systems which I very carefully tweaked to get the best image out of this wonderful source port.

Consider this, my current scenario: 3 WinXP-powered P4's, 2 of them outputting to 17-inch monitors, while the other one gets the privilege of a massive, high quality 21-incher, all of them CRT's. This setup works perfectly well for CD's 3 monitor mode, but there's a catch. One of these computers cannot use a resolution lower than 640x400, so I tell CD to upscale to 640x400, then I stretch the output 3:2 image to 4:3 from the monitor itself. The other computer cannot go lower than 640x480, so I created a custom powerstrip modeline that's essentially 640x480@70hz (original defaults to 60hz), then I tell CD to use that mode without aspect ratio correction on, and proceed also to stretch the 3:2 output image to 4:3, again, right from the monitor itself for a perfect, evenly-scaled, aspect ratio-corrected image. Then there's the last computer, the one with the 21-monitor, which's the only monitor that can handle 1600x1200. Problem is, that resolution puts way too much strain on the deflection yoke and the display mode itself I have not been able to tweak down to a more or less perfect 70hz, which's absolutely necessary when vsync's on (and you want that on if you don't like tearing), lest the left/right turning movement uncovers a nasty stuttering by reason of the discrepancy between the 35fps generated by Doom and the non-perfect 70hz mode which's roughly 1hz off, and because of this, I stick to good ole' 320x200 fullscreen mode on the 21-inch CRT setup, which looks beyond fantastic and can actually use my custom 70hz.

Now, after reading through the thread, there's no way 3:2 resolutions (not even the original 320x200, as it too would have to be subject to hardware rescaling) will pass the CRC32 test under the proposed approach, so that effectively renders everything I spent years setting up and tweaking for pixel accuracy's sake, completely and absolutely useless.

Consider thou the cry of thine servant.

I'm not even going to try to make some sort of fraudulent, emotional argument here by pointing out the "hundreds of thousands that will be disappointed when software scaling's no more", because it could very well be that I'm the only Chocolate Doom user that cares about this stuff, and what am I that you should hearken unto me?

Yet, if software scaling has somehow become so much of a burden to maintain that getting rid of it is the only reasonable way to perpetuate the existence of Chocolate Doom, then i suppose I could switch to DosBox, but the performance hit would probably smoke the other 2 computers out of existence, and quite frankly CD makes it all so easy, specially on the dehacked-networking departments, that it would be hard for me to go back to something more primitive, but what choice would I have if I want the real deal?

Nevertheless, it is my earnest desire for this project that the current software scaling be preserved intact as a fall-back, selectable option. And even if you don't see much of a use for it right now, it might prove a wiser choice in the long run, specially in light of the Chocolate Doom philosophy.

Best regards and sorry for the inconveniences.

**Comment by AXDOOMER** [comment]
The software rendered could be kept for legacy purposes. If you have a old Linux system, then you may not be able to play with the new hardware scaler because your OpenGL drivers won't work. I have a 2008 laptop with an ATI video card, some OpenGL game run, other don't even start. When they run, the only things that get drawn to the screen are empty blue and yellow shapes. 

Personally, I think there should be one last release of Chocolate-Doom with the software renderer before the hardware one becomes the only renderer and it should be available on Chocolate-Doom's website to download for each platform (including Linux .deb and .rpm files). 

People who don't want to or can't use the hardware renderer will download this one. Chocolate-Doom is very stable now, there are not a lot of features that are being added and it is trustworthy for testing vanilla-compatible maps. On my opinion, there's no need to keep the software code because it's huge and every resolution must be added manually into the code. But before the software mode is removed, the hardware scaler must replicate exactly what the software scaler is doing right now and that is scaling in one direction only so the interpolation is unapparent.

**Comment by AXDOOMER** [comment]
> For each, I took a screenshot at the actual resolution and then increased the window size by just a few pixels so it was almost the same but not exactly: this forces interpolation and allows us to compare the effect.

If you scale Doom's output by a tiny bit, this forces interpolation between almost (if not all) the pixels. The problem with this is that the interpolated pixel is closer to one pixel than the other, so it the resulting color looks closer to the closest parent pixel, which means the pixels look sharper. 

If you change the resolution dramatically, for example 1024x768, then the pixels that are the result of interpolation are push at look more towards the center of their parents. So they take more of the color of each and this makes the screen a lot more blurry:

https://i.imgur.com/6Y9JqMh.png

If you compare both screenshots, mine has a lot more horizontal interpolation, so it's a lot blurrier. 
(the resolution of the picture is 1023x768, but this was taken at 1024x768, I just accidentally chopped a column in Paint). It was taken in DosBox with scaling options that do exactly what you are trying to achieve with the hardware renderer.

**Comment by AXDOOMER** [comment]
Here's the difference. The picture is an analysis of the impact that a certain resolution can have on the interpolation, continuing my previous comment. 

@fragglet: Mine is left and yours is right. My resolution is a lot farther to a multiple of Doom's original resolution, so the interpolation is a lot more visible. In your screenshot, it was 2 pixels away so even if it created horizontal interpolation, the resulting pixels look more like their parent pixels, thus making it appear sharper. Mine's a lot blurrier because of the resolution I have chosen. 

http://i.imgur.com/0Bs8qL4.png

**Comment by AXDOOMER** [comment]
![30cqdc2](https://cloud.githubusercontent.com/assets/6194072/10445471/31fb2a46-7141-11e5-858f-f6a8948a2a2a.png)
This is a black and white picture from two pixels taken in Fraggle's screenshot. I'm just wondering if any of you can sees the interpolation. The interpolated pixel is different by only 1/255 (R, G and B), but I see it.

**Comment by fragglet** [comment]
@PorscheMonty I'm surprised by your comment because I've stated several times that everything in your setup should work at least as good as the current software renderer. In some cases it should actually be an improvement. No offence but it honestly seems like you've completely misunderstood everything I've said.

**Comment by fragglet** [comment]
@AXDOOMER - thanks for your comments. This is essentially the point I was making in my previous comment where I linked the gallery.

I have some more thoughts about this but I need time to do some research before they're fully formed and am ready to share them (I'm writing a script to collect some data and test some hypotheses I have). However, I would say that the conversation is in danger of heading in an unhelpful direction if we get bogged down in microscopic examinations of individual screen pixels. Of course interpolation will exist if you don't use an exact integer multiple - the real question is how significant this is. I put the gallery together so that we can evaluate the effect of non-integer scaling in a subjective manner.

But all of this is really a completely separate discussion to the one @PorscheMonty is bringing up as he's using integer scaling anyway. None of this should be an issue. I think there's some confusion here because two different conversations are taking place at once.

**Comment by fragglet** [comment]
@PorscheMonty - Once more to be completely explicit and just because I perhaps wasn't clear enough in communicating previously:

* The ability to run without the aspect ratio correction feature (ie. 320x200, 640x400, 960x600, etc.) will still be there.
* You should still be able to run 640x400 letterboxed in 640x480 mode. I would *expect* any screen modes supported by SDL1 to work in SDL2 too.
* 1600x1200 should be pixel identical to 1600x1200 with the current software scaler. There is no reason at all that it should be any different.
* The only thing I'm not sure about is vsync. SDL2 has different video backend code to SDL1. But SDL2 does have [this API function](https://wiki.libsdl.org/SDL_GL_SetSwapInterval) which potentially looks promising.

Most of the discussion here concerns the quality when using screen interpolation to run at resolutions that previously weren't supported (eg. 1920x1080). This should all be completely irrelevant to you because you're not using it anyway.

**Comment by PorscheMonty** [comment]
@fragglet: I believe it's the "improvement" part you keep bringing up what actually got me confused, given I understood it within the context of picture quality on my current setup (which cannot be improved upon as its already perfect) rather than potential performance gain, which seems to be what you're actually suggesting here.

It all sounded to me like "it won't look exactly the same but you'll be pleased with the results anyways".

Sorry about that. I tend to get carried away with details.
Regards.

**Comment by fragglet** [comment]
Okay, understood. Apologies for the confusion.

**Comment by fabiangreffrath** [comment]
@PorscheMonty Let me add that the hardware-scaling branch will never *change* your video mode. If you select a fullscreen mode, it will run with whatever resolution your desktop has and scale the game screen accordingly. So, if you can set your desktop to 640x400 and have your monitor stretch it to 4:3, the hardware-scaling branch will run without any interpolation.

**Comment by PorscheMonty** [comment]
Thanks, and sorry about all that junk. I do certainly appreciate the extra information.

**Comment by jmtd** [comment]
Belated happy new year, bug!

I've picked up looking at c-d again, and I'm starting with input events. in the setup programs, any of the inputs that are remapped (e.g. "a" for advanced, etc.) don't work. Poking at the source, this is only a problem if key_mapping is on (bodge the default to off and things work OK). Investigations continue...

**Comment by fragglet** [comment]
Input stuff should be fixed now, so if it still doesn't work then it's news to me. It's only libtextscreen that has its inputs broken.

**Comment by fragglet** [comment]
I think the conclusion I reached was that we needed three input modes for libtextscreen and the ability to switch between them:

* "Normal" mode where we generate text input based on key events
* "Input" mode where we generate text input based on input events (and calling SDL_StartTextInput() to pop up the on-screen keyboard if necessary).
* "Key" mode where we do the same mapping as i_video.c, for use when setting key bindings.

**Comment by jmtd** [comment]
OK, I've read through the i_video stuff, and I'm not quite as up to speed as you must be with regards this, but things don't seem quite right to me. [This patch](https://github.com/jmtd/chocolate-doom/commit/14801c002ccd88b5e48e3eafaea1e059c04ec928) is sufficient to fix `chocolate-doom-setup` for me, but I have a feeling more work is needed. I haven't experimented much yet with typing unicode chars, but things don't seem quite right: I did try to set a key to "€", the setup program looked like the attached picture, but I couldn't verify that it worked in `chocolate-doom` (too hard to type!)

![screen shot 2016-03-02 at 21 09 32](https://cloud.githubusercontent.com/assets/63694/13475487/191d6d56-e0bb-11e5-9e37-547b6fb5f02e.png)

I couldn't provide § as a symbol, probably because it's one of the hard-coded keys? (as `)

**Comment by jmtd** [comment]
What's left to fix for sdl2-branch - am I missing anything?

* ☑ ~~video~~ - is this now done?
  * ☐ HOM emulation - review #676 
    * ☐ STDISK drawing problems - review #668 - depends on #676 
* ☑ ~~text input~~ - done for engines?
 * ☐ libtextscreen (discussion in this bug)
* ☐ joystick input (#673, possibly also addressing #653 )

I'd like to help on some of this. Unless anyone objects, I'd probably focus on libtextscreen initially.

**Comment by fragglet** [comment]
See here for the list of things to do:

https://www.chocolate-doom.org/wiki/index.php/Sdl2-branch

(Feel free to add joystick input to the list)

**Comment by fragglet** [comment]
This is too big for one bug; I'm closing this and recommending people follow along on the milestone:

https://github.com/chocolate-doom/chocolate-doom/milestones/Merge%20SDL2%20branch
